r/R/trajectory_analysis.R

```diff
@@ -0,0 +1,68 @@
+runTrajectoryAnalysis <- function(req, data) {
+  root_nodes <- req$body$rootNodes
+
+  cell_data <- generateGraphData(data)
+  cell_data <- monocle3::order_cells(cell_data, reduction_method = "UMAP", root_pr_nodes = root_nodes)
+
+  # monocle3::plot_cells(cell_data,color_cells_by = "pseudotime",label_cell_groups=FALSE,label_leaves=FALSE,label_branch_points=FALSE,graph_label_size=1.5)
+
+  pseudotime <- as.data.frame(cell_data@principal_graph_aux@listData$UMAP$pseudotime)
+  pseudotime$cells_id <- data@meta.data$cells_id
+  pseudotime <- pseudotime[order(pseudotime$cells_id), ]
+  pseudotime <- pseudotime %>%
+    tidyr::complete(cells_id = seq(0, max(data@meta.data$cells_id))) %>%
+    dplyr::select(-cells_id)
+  return(unname(pseudotime))
+}
+
+runGenerateTrajectoryGraph <- function(req, data) {
+  cell_data <- generateGraphData(data)
+  node_coords <- t(cell_data@principal_graph_aux[["UMAP"]]$dp_mst)
+  umap_coords <- as.data.frame(SingleCellExperiment::reducedDims(cell_data)[["UMAP"]])
+
+  # node coordinates
+  # get connected nodes
+  connected_nodes <- list()
+  for (node in rownames(node_coords)) {
+    connected_nodes[[node]] <- as.list(names(cell_data@principal_graph[["UMAP"]][[which(rownames(node_coords) == node)]][[1]]))
```

Comment:
this could be made clearer, it's super difficult to read. for starters, extract the which(rownames(node_coords) == node) to an aptly named variable.